File: src/lib/commands/graph.ts
import { invoke } from '@tauri-apps/api/core';
import type { Mash } from '$lib/types';

export interface GraphNode {
	id: string;
	mashType: string;
	summary: string;
	context: string;
	memo: string;
	status: string;
	createdAt: number;
	updatedAt: number;
}

export interface GraphEdge {
	id: number;
	sourceId: string;
	targetId: string;
	relationType: string;
	source: string;
	confidence: number;
	createdAt: number;
	updatedAt: number;
}

export interface GraphData {
	nodes: GraphNode[];
	edges: GraphEdge[];
}

export interface GraphFilters {
	mashTypes?: string[];
	relationTypes?: string[];
	sources?: string[];
}

export function getGraph(filters: GraphFilters): Promise<GraphData> {
	return invoke<GraphData>('get_graph', { filters });
}

export function getNodeDetail(id: string): Promise<GraphData> {
	return invoke<GraphData>('get_node_detail', { id });
}

export function expandNode(id: string, depth: number): Promise<GraphData> {
	return invoke<GraphData>('expand_node', { id, depth });
}

export function addEdge(
	sourceId: string,
	targetId: string,
	relationType: string,
	source: string,
	confidence: number
): Promise<GraphEdge> {
	return invoke<GraphEdge>('add_edge', { sourceId, targetId, relationType, source, confidence });
}

export function updateEdge(id: number, relationType?: string, confidence?: number): Promise<GraphEdge> {
	return invoke<GraphEdge>('update_edge', { id, relationType, confidence });
}

export function deleteEdge(id: number): Promise<void> {
	return invoke('delete_edge', { id });
}

export function updateNode(
	id: string,
	mashType?: string,
	summary?: string,
	context?: string,
	memo?: string
): Promise<Mash> {
	return invoke<Mash>('update_node', { id, mashType, summary, context, memo });
}

File: src-tauri/src/commands/graph.rs
use tauri::State;

use crate::commands::DbState;
use crate::db;
use crate::models::{Edge, GraphData, GraphFilters};

#[tauri::command]
pub fn get_graph(
    state: State<DbState>,
    filters: GraphFilters,
) -> Result<GraphData, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::edges::get_graph(&conn, &filters)
}

#[tauri::command]
pub fn get_node_detail(
    state: State<DbState>,
    id: String,
) -> Result<GraphData, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::edges::get_node_neighbors(&conn, &id)
}

#[tauri::command]
pub fn expand_node(
    state: State<DbState>,
    id: String,
    depth: u32,
) -> Result<GraphData, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::edges::expand_node(&conn, &id, depth)
}

#[tauri::command]
pub fn add_edge(
    state: State<DbState>,
    source_id: String,
    target_id: String,
    relation_type: String,
    source: String,
    confidence: f64,
) -> Result<Edge, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::edges::add_edge(&conn, &source_id, &target_id, &relation_type, &source, confidence)
}

#[tauri::command]
pub fn update_edge(
    state: State<DbState>,
    id: i64,
    relation_type: Option<String>,
    confidence: Option<f64>,
) -> Result<Edge, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::edges::update_edge(&conn, id, relation_type.as_deref(), confidence)
}

#[tauri::command]
pub fn delete_edge(
    state: State<DbState>,
    id: i64,
) -> Result<(), String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::edges::delete_edge(&conn, id)
}

#[tauri::command]
pub fn update_node(
    state: State<DbState>,
    id: String,
    mash_type: Option<String>,
    summary: Option<String>,
    context: Option<String>,
    memo: Option<String>,
) -> Result<crate::models::Mash, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    db::mashes::update_mash(
        &conn,
        &id,
        mash_type.as_deref(),
        summary.as_deref(),
        context.as_deref(),
        memo.as_deref(),
    )
}
